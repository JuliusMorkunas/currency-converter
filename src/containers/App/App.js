import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { useDebouncedCallback } from 'use-debounce';
import s from './App.module.scss';
import {
  initialize,
  onChangeAmountFrom,
  onChangeAmountTo,
  toggleCurrencyFrom,
  toggleCurrencyTo,
  switchCurrencies,
  selectCurrencyFrom,
  selectCurrencyTo,
  getRate,
} from '../../redux/modules/converter';
import CurrencySelector from '../../components/CurrencySelector';
import Switch from '../../components/Switch';
import AmountInput from '../../components/AmountInput';
import Disclaimer from '../../components/Disclaimer';
import ExchangeRate from '../../components/ExchangeRate';
import Button from '../../components/Button';
import { formatToCurrency } from '../../utils/helpers';

const LABELS = {
  from: 'From',
  to: 'To',
  amount: 'Amount',
  converted: 'Converted to',
};
const INPUT_DEBOUNCE_MS = 500;
const ENTER_KEY_CODE = 13;

export function App({
  currencies,
  currencyFrom,
  currencyTo,
  isOpenFrom,
  isOpenTo,
  isInitialized,
  amountFrom,
  amountTo,
  rate,
  initialize,
  onChangeAmountFrom,
  onChangeAmountTo,
  toggleCurrencyFrom,
  toggleCurrencyTo,
  switchCurrencies,
  selectCurrencyFrom,
  selectCurrencyTo,
  getRate,
}) {
  // TODO: Might want to use exact calculations from API instead of doing math on the client side based on fetched rate
  const [debouncedOnChangeAmountFrom] = useDebouncedCallback(async value => {
    const rate = await getRate();
    onChangeAmountTo(formatToCurrency(rate ? value * rate : 0));
  }, INPUT_DEBOUNCE_MS);
  const [debouncedOnChangeAmountTo] = useDebouncedCallback(async value => {
    const rate = await getRate();
    onChangeAmountFrom(formatToCurrency(rate ? value / rate : 0));
  }, INPUT_DEBOUNCE_MS);

  const onInitialInputKeyDown = e => {
    if (e.keyCode === ENTER_KEY_CODE) {
      initialize();
    }
  };

  useEffect(() => {
    getRate && getRate();
  }, [getRate]);

  return (
    <div className={s.Container}>
      <div className={s.CurrencySelectorsContainer}>
        <CurrencySelector
          currencies={currencies}
          currency={currencyFrom}
          label={LABELS.from}
          isOpen={isOpenFrom}
          onToggle={toggleCurrencyFrom}
          onSelect={selectCurrencyFrom}
        />
        <Switch onClick={switchCurrencies} />
        <CurrencySelector
          currencies={currencies}
          currency={currencyTo}
          label={LABELS.to}
          isOpen={isOpenTo}
          onToggle={toggleCurrencyTo}
          onSelect={selectCurrencyTo}
        />
      </div>
      {isInitialized ? (
        <>
          <div className={s.AmountInputsContainer}>
            <AmountInput
              label={LABELS.amount}
              currency={currencyFrom}
              value={amountFrom}
              onChange={({ target: { value } }) => {
                if (isNaN(value)) {
                  return;
                }
                onChangeAmountFrom(value);
                debouncedOnChangeAmountFrom(value);
              }}
            />
            <AmountInput
              label={LABELS.converted}
              currency={currencyTo}
              value={amountTo}
              onChange={({ target: { value } }) => {
                if (isNaN(value)) {
                  return;
                }
                onChangeAmountTo(value);
                debouncedOnChangeAmountTo(value);
              }}
            />
          </div>
          <ExchangeRate currencyFrom={currencyFrom} currencyTo={currencyTo} rate={rate} />
          <Disclaimer>
            All figures are live mid-market rates, which are for informational purposes only. To see
            the rates for money transfer, please select sending money option.
          </Disclaimer>
        </>
      ) : (
        <>
          <div className={s.InitialAmountInputContainer}>
            <AmountInput
              label={LABELS.amount}
              currency={currencyFrom}
              value={amountFrom}
              onKeyDown={onInitialInputKeyDown}
              onChange={({ target: { value } }) => {
                if (isNaN(value)) {
                  return;
                }
                onChangeAmountFrom(value);
                debouncedOnChangeAmountFrom(value);
              }}
              autoFocus
            />
          </div>
          <Button onClick={initialize}>Convert</Button>
        </>
      )}
    </div>
  );
}

const ConnectedApp = connect(
  ({ converter }) => ({ ...converter }),
  {
    initialize,
    onChangeAmountFrom,
    onChangeAmountTo,
    toggleCurrencyFrom,
    toggleCurrencyTo,
    switchCurrencies,
    selectCurrencyFrom,
    selectCurrencyTo,
    getRate,
  },
)(App);

App.propTypes = {
  currencies: PropTypes.array.isRequired,
  currencyFrom: PropTypes.string.isRequired,
  currencyTo: PropTypes.string.isRequired,
  isOpenFrom: PropTypes.bool.isRequired,
  isOpenTo: PropTypes.bool.isRequired,
  isInitialized: PropTypes.bool.isRequired,
  amountFrom: PropTypes.string.isRequired,
  amountTo: PropTypes.string.isRequired,
  rate: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  initialize: PropTypes.func.isRequired,
  onChangeAmountFrom: PropTypes.func.isRequired,
  onChangeAmountTo: PropTypes.func.isRequired,
  toggleCurrencyFrom: PropTypes.func.isRequired,
  toggleCurrencyTo: PropTypes.func.isRequired,
  switchCurrencies: PropTypes.func.isRequired,
  selectCurrencyFrom: PropTypes.func.isRequired,
  selectCurrencyTo: PropTypes.func.isRequired,
  getRate: PropTypes.func.isRequired,
};

export default ConnectedApp;
