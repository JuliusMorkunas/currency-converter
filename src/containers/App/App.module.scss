@import '../../variables';

.Container {
  font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Helvetica Neue, Arial,
    sans-serif;
  display: flex;
  flex-direction: column;
  box-sizing: border-box;
  width: 564px;
  padding: 40px 52px;
  border-radius: 1px;
  background-color: $c-white;
  box-shadow: 0 0 15px 0 $c-bluish-shadow;

  @media (max-width: $b-small) {
    width: 100%;
    padding: 20px;
  }

  * {
    box-sizing: border-box;
  }
}

.CurrencySelectorsContainer {
  display: flex;
  align-items: flex-end;
  justify-content: space-between;
  margin-bottom: 30px;

  > div {
    width: 42%;
  }
}

.InitialAmountInputContainer {
  margin-bottom: 72px;
}

.AmountInputsContainer {
  display: flex;
  justify-content: space-between;
  margin-bottom: 32px;

  @media (max-width: $b-small) {
    margin-bottom: 16px;
  }

  > div {
    width: 42%;
  }
}
