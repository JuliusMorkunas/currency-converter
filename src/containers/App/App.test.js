import React from 'react';
import ReactDOM from 'react-dom';
import { render } from '@testing-library/react';
import { App } from './App';

const props = {
  currencies: ['eur', 'gbp'],
  currencyFrom: 'eur',
  currencyTo: 'gbp',
  isOpenFrom: false,
  isOpenTo: false,
  isInitialized: false,
  amountFrom: '0',
  amountTo: '0',
  rate: '1',
  initialize: () => {},
  onChangeAmountFrom: () => {},
  onChangeAmountTo: () => {},
  toggleCurrencyFrom: () => {},
  toggleCurrencyTo: () => {},
  switchCurrencies: () => {},
  selectCurrencyFrom: () => {},
  selectCurrencyTo: () => {},
  getRate: () => {},
};

describe('Uninitialized App', () => {
  it('should render without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App {...props} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('should render correctly', () => {
    const { asFragment } = render(<App {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Uninitialized App', () => {
  it('should render correctly', () => {
    const { asFragment } = render(<App {...props} isInitialized={true} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
