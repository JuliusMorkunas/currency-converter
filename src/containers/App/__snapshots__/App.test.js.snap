// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Uninitialized App should render correctly 1`] = `
<DocumentFragment>
  <div
    class="Container"
  >
    <div
      class="CurrencySelectorsContainer"
    >
      <div
        class="Container "
      >
        <label
          class="Label"
        >
          From:
        </label>
        <div
          class="InputContainer"
        >
          <div
            class="SelectedCurrencyContainer"
          >
            <div
              class="CurrencyIcon eur"
            />
            <div
              class="CurrencyName"
            >
              eur
            </div>
          </div>
          <span
            class="DropdownArrow"
          />
        </div>
      </div>
      <button
        class="Button"
      >
        <span
          class="Icon"
        />
      </button>
      <div
        class="Container "
      >
        <label
          class="Label"
        >
          To:
        </label>
        <div
          class="InputContainer"
        >
          <div
            class="SelectedCurrencyContainer"
          >
            <div
              class="CurrencyIcon gbp"
            />
            <div
              class="CurrencyName"
            >
              gbp
            </div>
          </div>
          <span
            class="DropdownArrow"
          />
        </div>
      </div>
    </div>
    <div
      class="InitialAmountInputContainer"
    >
      <div
        class="Container"
      >
        <label
          class="Label"
          for="amount-amount"
        >
          Amount:
        </label>
        <input
          class="Input"
          id="amount-amount"
          name="amount-amount"
          type="text"
          value="0"
        />
        <span
          class="Currency"
        >
          eur
        </span>
      </div>
    </div>
    <button
      class="Button"
    >
      Convert
    </button>
  </div>
</DocumentFragment>
`;

exports[`Uninitialized App should render correctly 2`] = `
<DocumentFragment>
  <div
    class="Container"
  >
    <div
      class="CurrencySelectorsContainer"
    >
      <div
        class="Container "
      >
        <label
          class="Label"
        >
          From:
        </label>
        <div
          class="InputContainer"
        >
          <div
            class="SelectedCurrencyContainer"
          >
            <div
              class="CurrencyIcon eur"
            />
            <div
              class="CurrencyName"
            >
              eur
            </div>
          </div>
          <span
            class="DropdownArrow"
          />
        </div>
      </div>
      <button
        class="Button"
      >
        <span
          class="Icon"
        />
      </button>
      <div
        class="Container "
      >
        <label
          class="Label"
        >
          To:
        </label>
        <div
          class="InputContainer"
        >
          <div
            class="SelectedCurrencyContainer"
          >
            <div
              class="CurrencyIcon gbp"
            />
            <div
              class="CurrencyName"
            >
              gbp
            </div>
          </div>
          <span
            class="DropdownArrow"
          />
        </div>
      </div>
    </div>
    <div
      class="AmountInputsContainer"
    >
      <div
        class="Container"
      >
        <label
          class="Label"
          for="amount-amount"
        >
          Amount:
        </label>
        <input
          class="Input"
          id="amount-amount"
          name="amount-amount"
          type="text"
          value="0"
        />
        <span
          class="Currency"
        >
          eur
        </span>
      </div>
      <div
        class="Container"
      >
        <label
          class="Label"
          for="amount-convertedto"
        >
          Converted to:
        </label>
        <input
          class="Input"
          id="amount-convertedto"
          name="amount-convertedto"
          type="text"
          value="0"
        />
        <span
          class="Currency"
        >
          gbp
        </span>
      </div>
    </div>
    <div
      class="Container"
    >
      <span
        class="Decoration"
      />
      <span
        class="ExchangeRate"
      >
        1 eur = 1 gbp
      </span>
    </div>
    <p
      class="Disclaimer"
    >
      All figures are live mid-market rates, which are for informational purposes only. To see the rates for money transfer, please select sending money option.
    </p>
  </div>
</DocumentFragment>
`;
